src/profile-logic/marker-data.js

```diff
@@ -373,6 +375,232 @@ export function deriveMarkersFromRawMarkerTable(
+  rangeStart: number,
+  rangeEnd: number
+): RawMarkerTable {
+  const newMarkerTable = getEmptyRawMarkerTable();
+
+  const isTimeInRange = (time: number): boolean =>
+    time < rangeEnd && time >= rangeStart;
+  const intersectsRange = (start: number, end: number): boolean =>
+    start < rangeEnd && end >= rangeStart;
+  const copyIndexToNewTable = (i: IndexIntoRawMarkerTable) => {
+    newMarkerTable.time.push(markers.time[i]);
+    newMarkerTable.name.push(markers.name[i]);
+    newMarkerTable.data.push(markers.data[i]);
+    newMarkerTable.length++;
+  };
+
+  // These maps contain the start markers we find while looping the marker
+  // table.
+  // The first map contains the start markers for tracing markers. They can be
+  // nested and that's why we use an array structure as value.
+  const openTracingMarkers: Map<
+    IndexIntoStringTable,
+    IndexIntoMarkers[]
+  > = new Map();
+
+  // The second map contains the start markers for network markers.
+  // Note that we don't have more than 2 network markers with the same name as
+  // the name contains an incremented index. Therefore we don't need to use an
+  // array as value like for tracing markers.
+  const openNetworkMarkers: Map<
+    IndexIntoStringTable,
+    IndexIntoMarkers
+  > = new Map();
+
+  let previousScreenshotMarker = null;
+
+  for (let i = 0; i < markers.length; i++) {
+    const name = markers.name[i];
+    const time = markers.time[i];
+    const data = markers.data[i];
+
+    const addCurrentMarkerIfInRange = () => {
+      if (isTimeInRange(time)) {
+        copyIndexToNewTable(i);
+      }
+    };
+
+    if (!data) {
+      addCurrentMarkerIfInRange();
+      continue;
+    }
+
+    // Depending on the type we have to do some special handling.
+    switch (data.type) {
+      case 'tracing': {
+        // Tracing markers are pairs of start/end markers. To retain their
+        // duration if we have it, we keep both markers of the pair if they
+        // represent a marker that's partially in the range.
+
+        if (data.interval === 'start') {
+          let openMarkersForName = openTracingMarkers.get(name);
+          if (!openMarkersForName) {
+            openMarkersForName = [];
+            openTracingMarkers.set(name, openMarkersForName);
+          }
+          openMarkersForName.push(i);
+
+          // We're not inserting anything to newMarkerTable yet. We wait for the
+          // end marker to decide whether we should add this start marker, as we
+          // will add start markers from before the range if the end marker is
+          // in or after the range.
+          //
+          // We'll loop at all open markers after the main loop, to add them to
+          // the new marker table if they're in the range.
+        } else if (data.interval === 'end') {
+          const openMarkersForName = openTracingMarkers.get(name);
+          let startIndex;
+          if (openMarkersForName) {
+            startIndex = openMarkersForName.pop();
```

Comment:
This is a very good question!

This is the assumption we (I :) ) already used in https://github.com/firefox-devtools/profiler/blob/723f25a0f5c8d7cd42bcd16824e0da4626c6f021/src/profile-logic/marker-data.js#L276-L279

Ideally we'd have an id to match the start and end markers, but we don't have this information so we have to assume "something". This is what made the most sense to me at the time, and I think this is still the case (think DOMEvents). I'd be happy to change it in both place if you think this isn't right though, or if you have an example where this logic fails.

Reply:
I can't think of an example currently but for example in our tests, there are nested tracing markers in the code below. A test(has `should handle nested markers correctly` title) fails when we change this logic. I'm not sure markers can be nested.
https://github.com/firefox-devtools/profiler/blob/d5d3c9e7483c2fd70c70fae3593547a333fc2252/src/test/fixtures/profiles/gecko-profile.js#L321-L391

It feels like we should match the tracing markers like a queue instead of a stack.
But if the current logic uses that algorithm already, we can land this and file an issue to resolve this. (Also prefer to ask Markus about his opinions in there.)